GPUTreeShap/gpu_treeshap.h

```diff
@@ -28,20 +28,111 @@
 #include <vector>
 
 namespace gpu_treeshap {
+template <typename Return>
+class DiscardOverload : public thrust::discard_iterator<Return> {
+ public:
+  using value_type = Return;  // NOLINT
+};
+
+struct Bitset {
+  struct Span {
+    uint32_t *bits {nullptr};
+    size_t n {0};
+
+    __device__ __host__ size_t size() const { return n; }
+    __device__ __host__ uint32_t *data() const { return bits; }
+
+    __device__ __host__ uint32_t const &operator[](size_t i) const {
+      return bits[i];
+    }
+    __device__ __host__ uint32_t &operator[](size_t i) { return bits[i]; }
+    __device__ __host__ bool empty() const { return size() == 0; }
+  } bits;
+
+  uint32_t static constexpr kOne = 1;
+  static uint32_t constexpr kValueSize = sizeof(uint32_t) * 8;
+
+  Bitset() = default;
+  __device__ __host__ Bitset(uint32_t *ptr, size_t n) : bits{ptr, n} {}
+
+  struct Pos {
+    uint32_t int_pos{0};
+    uint32_t bit_pos{0};
+  };
+
+  __device__ __host__ static Bitset Intersect(Bitset l, Bitset r) {
+    if (l.bits.data() == r.bits.data()) {
+      return l;
+    }
+    if (r.bits.size() > l.bits.size()) {
+      thrust::swap(l, r);
+    }
+    for (size_t i = 0; i < r.bits.size(); ++i) {
+      l.bits[i] &= r.bits[i];
+    }
+    return l;
+  }
+
+  __device__ __host__ static Pos ToBitPos(uint32_t pos) {
+    Pos pos_v;
+    if (pos == 0) {
+      return pos_v;
+    }
+    pos_v.int_pos = pos / kValueSize;
+    pos_v.bit_pos = pos % kValueSize;
+    return pos_v;
+  }
+
+  __device__  __host__ Pos static Shift(Pos pos) {
+    pos.bit_pos = sizeof(uint32_t) * 8 - pos.bit_pos - uint32_t(1);
+    return pos;
+  }
+
+  __device__ __host__ bool Check(Pos pos_v) const {
+    pos_v = Shift(pos_v);
+    uint32_t const value = bits[pos_v.int_pos];
+    uint32_t const test_bit = kOne << pos_v.bit_pos;
+    uint32_t result = test_bit & value;
+    return static_cast<bool>(result);
+  }
+
+  __device__ __host__ bool Check(uint32_t bit) const {
+    Pos pos_v = ToBitPos(bit);
+    return Check(pos_v);
+  }
+
+  __device__ __host__ void Set(uint32_t pos) {
+    Pos pos_v = Shift(ToBitPos(pos));
+    uint32_t& value = bits[pos_v.int_pos];
+    uint32_t set_bit = kOne << pos_v.bit_pos;
+    value |= set_bit;
+  }
+};
 
 struct XgboostSplitCondition {
```

Comment:
Because PathElement accepts a split condition as a template, I think you can just put the custom split condition implementation inside xgboost. That would mean you don't have to implement span/bitset here.

Reply:
Moved the new implementation back to XGBoost.  This one is kept unchanged for benchmarks and tests.